Extract migration queue from InternalPartitionServiceImpl

metanet opened the pull request:
Extract management of migrations tasks into a separate class. It makes it easier to manage migration state of the partition service. It also solves the race problem occurs during starting a migration task polled from the migration queue. 

Fixes #7191

metanet commented:
run-lab-run

mdogan commented:
:+1:

pveentjer commented on hazelcast/src/main/java/com/hazelcast/partition/impl/InternalPartitionServiceImpl.java:
```diff
@@ -2102,15 +2102,22 @@ public void run() {
+
+                Runnable r = migrationQueue.peek();
                 if (r == null) {
+                    Thread.sleep(100);
```
We are not checking 10 more frequent than before? Because initially the frequency was 1/second and now 10/second.

pveentjer replied:
@metanet ping.

metanet replied:
Before this change, ``poll()`` is used with 1 second of timeout. So if there is no migration task, it will be checked every 1 second. Once a migration task is put into the queue, it will be polled immediately. I used 100 ms sleep duration to get the migration task without too much delay.

pveentjer commented:
Could you add a short explanation of the race?

metanet commented:
MigrationThread polls a migration task from migrationQueue and sets the migrating flag but this is not an atomic operation. hasOnGoingMigrationLocal() returns false when a migration task is polled from the queue but the flag is not set yet. This fix first sets the flag before polling the task from the queue.

metanet commented:
@pveentjer pr description is updated

pveentjer commented:
Sorry for my stupidity; so apparently another thread than the MigrationThread is suffering from this non atomic behavior. And this thread makes the wrong conclusion; it thinks there are no migrations since the flag isn't set, but actually there are migrations. Is my understanding correct?

Which thread is this other thread?
What are the consequences of this race?

metanet commented:
Any code that checks about the safety of the partitions may think that there is no migration while there is a migration task to be runned. For example, instance shut down.

pveentjer commented:
But if the calling code would have called e.g. just before the MigrationThread would have done its work, it would also see a 'stale' value since the flag hasn't been set. So why is it important that the other threads needs to see 'the most actual' value?

metanet commented:
There are 2 cases:
1) There is migration task in the queue
2) Migration task is polled from the queue and being run

For both of these cases, ``hasOnGoingMigrationLocal()``, which can be called by any other thread,must return true. Between step 1 and step 2, there is a race which is fixed by this test.

pveentjer commented:
Sorry for my never ending stupidity. Would it not make sense to have an AtomicLong that is incremented with every migration task and decremented when a tasks is completed? To check if there are ongoing migration, check if the value > 0. 

My lack of understanding the threading for this particular section is causing problems.

metanet commented:
Np. That was our other solution but it creates its own problems because for some parts of the codes, migrationQueue is updated within a lock; for some other parts, it is updated without a lock. 

We will refactor these parts while working on partition service for 3.7

pveentjer commented:
So something like this:

'''
 private class MigrationThread extends Thread implements Runnable {
        private final long sleepTime = max(250L, partitionMigrationInterval);
        private final AtomicLong migrationTaskCount = new AtomicLong();

        MigrationThread(Node node) {
            super(node.getHazelcastThreadGroup().getInternalThreadGroup(),
                    node.getHazelcastThreadGroup().getThreadNamePrefix("migration"));
        }

        @Override
        public void run() {
            try {
                while (!isInterrupted()) {
                    doRun();
                }
            } catch (InterruptedException e) {
                if (logger.isFinestEnabled()) {
                    logger.finest("MigrationThread is interrupted: " + e.getMessage());
                }
            } catch (OutOfMemoryError e) {
                OutOfMemoryErrorDispatcher.onOutOfMemory(e);
            } finally {
                migrationQueue.clear();
            }
        }
        
        private void doRun() throws InterruptedException {
            for (; ; ) {
                if (!isMigrationAllowed()) {
                    break;
                }
                Runnable r = migrationQueue.poll(1, TimeUnit.SECONDS);
                if (r == null) {
                    break;
                }

                processTask(r);
                if (partitionMigrationInterval > 0) {
                    Thread.sleep(partitionMigrationInterval);
                }
            }
            boolean hasNoTasks = migrationQueue.isEmpty();
            if (hasNoTasks) {
                if (migrating) {
                    migrating = false;
                    logger.info("All migration tasks have been completed, queues are empty.");
                }
                evictCompletedMigrations();
                Thread.sleep(sleepTime);
            } else if (!isMigrationAllowed()) {
                Thread.sleep(sleepTime);
            }
        }
        private void postTask(Runnable task){
            if(task instanceof MigrateTask){
                migrationTaskCount.incrementAndGet();
            }
            migrationQueue.add(task);
        }
        boolean processTask(Runnable r) {
            if (r == null || isInterrupted()) {
                return false;
            }

            try {
                r.run();
            } catch (Throwable t) {
                logger.warning(t);
            }

            if(r instanceof MigrateTask){
                migrationTaskCount.decrementAndGet();
            }

            return true;
        }

        void stopNow() {
            migrationQueue.clear();
            interrupt();
        }

        boolean isMigrating() {
            return migrationTaskCount.get()>0;
        }
    }
'''

pveentjer commented:
After a short discussion with Basri, the solution with the counter makes sense to some point. However there is a lot of complex queue interaction like clear which could leave the counter in a permanently inconsistent state. I'll do some more thinking.

pveentjer commented:
@metanet I currently have a bunch of other tasks on my plate. Please have a look at the following code that should simplify the logic.

```
 static class MigrationQueue {
        private final AtomicInteger migrationTaskCount = new AtomicInteger();
        private final BlockingQueue<Runnable> taskQueue = new LinkedBlockingQueue<Runnable>();

        public Runnable poll(int timeout, TimeUnit unit) throws InterruptedException {
            return taskQueue.poll(timeout, unit);
        }

        public void add(Runnable task){
            if(task instanceof MigrateTask){
                migrationTaskCount.incrementAndGet();
                taskQueue.add(new MigrationTaskWrapper((MigrateTask)task));
            }else{
                taskQueue.add(task);
            }
        }

        public void clear() {
            List<Runnable> sink = new LinkedList<Runnable>();
            taskQueue.drainTo(sink);

            for(Runnable task: sink){
                if(task instanceof MigrateTask){
                    migrationTaskCount.decrementAndGet();
                }
            }
        }

        public boolean hasMigrations(){
            return migrationTaskCount.get()>0;
        }

        public boolean isEmpty() {
            return taskQueue.isEmpty();
        }

        public int size() {
            return taskQueue.size();
        }

         class MigrationTaskWrapper implements Runnable{
            private final MigrateTask task;

            public MigrationTaskWrapper(MigrateTask task) {
                this.task = task;
            }

            @Override
            public void run() {
                try {
                    task.run();
                }finally {
                    migrationTaskCount.decrementAndGet();
                }
            }
        }
    }
```

metanet commented:
@pveentjer pr is updated

pveentjer commented on hazelcast/src/main/java/com/hazelcast/partition/impl/InternalPartitionServiceImpl.java:
```diff
@@ -1254,8 +1252,8 @@ private boolean hasOnGoingMigrationMaster(Level level) {
-        return !activeMigrations.isEmpty() || !migrationQueue.isEmpty()
-                || migrationThread.isMigrating();
+        return !activeMigrations.isEmpty() || migrationQueue.isNonEmpty()
+                || migrationQueue.hasMigrations();
```
why the complex query? The MigrationQueue knows if there is a MigrationTask pending or executing. So we does activeMigrations and migrationQueue no empty be used?

So why not just:

```
 public boolean hasOnGoingMigrationLocal() {return migrationQueue.hasMigrations();}
```

mdogan replied:
Only master node knows `migrationQueue`, on other nodes it's always empty. `activeMigrations ` check is required for non-master nodes.

pveentjer replied:
Probably it would make sense to add some comments to explain to for others.

pveentjer commented on hazelcast/src/main/java/com/hazelcast/partition/impl/InternalPartitionServiceImpl.java:
```diff
@@ -2130,15 +2124,18 @@ private void doRun() throws InterruptedException {
             } catch (Throwable t) {
                 logger.warning(t);
+            } finally {
+                migrationQueue.afterTaskCompletion(r);
```
Why did you remove the wrapper task? With the wrapper task all this stuff remains an implementation detail of the queue. Now you have added the complexity to the thread.

mdogan replied:
`migrationQueue` itself is also a detail, it's directly bound to the partition service. Moving complexity from one side to another doesn't make any difference.

pveentjer replied:
It does. Now you have a mo ecomplex method than needs to be. With the wrapper task approach I have given, this is fully an implementation detail of the MigrationQueue. The logic for this thread is simplified since it doesn't need to do any callbacks to the queue. In my book this is cleaner code since incrementing and decrementing the counter is a implementation detail of the MigrationQueue.

mdogan replied:
If we start to talk things in our books then I think there's no need to discuss about this PR anymore. Discussion here started to become about esthetics and choices, again and again.

metanet replied:
Partition service already manages the whole migration process. Migration queue is only responsible for managing migration tasks. Therefore, having an extra callback method doesn't seem problematic to me. The other approach, wrapping the given runnable, is also ok but I don't think it is necessary and I want to keep our current approach.

pveentjer commented on hazelcast/src/test/java/com/hazelcast/partition/impl/MigrationQueueTest.java:
```diff
@@ -0,0 +1,77 @@
+    private final MigrationQueue migrationQueue = new MigrationQueue();
+
+    @Test
+    public void test_migrationTaskCount_incremented_onMigrateTask() {
```
Why does every test have 'test_migrationTaskCount' as prefix? It doesn't seem to add a lot of additional info,

metanet replied:
It is just my preference

pveentjer replied:
Do you think it is wise if we allow everybody to have their own preference when writing a test? Imho it makes code harder to understand. 

We have a test practices page on the wiki where we have a naming approach. If you don't like it and want to improve it, I'm all ear. But imho we should do this consistently using the whole system. I

pveentjer commented on hazelcast/src/test/java/com/hazelcast/partition/impl/MigrationQueueTest.java:
```diff
@@ -0,0 +1,77 @@
+    @Test
+    public void test_migrateTaskCount_decremented_afterMigrateTaskCompleted()
+            throws InterruptedException {
+        final InternalPartitionServiceImpl.MigrateTask migrateTask = mock(InternalPartitionServiceImpl.MigrateTask.class);
```
Can you make this a static import?

metanet replied:
It is a non static inner class

pveentjer commented on hazelcast/src/test/java/com/hazelcast/partition/impl/MigrationQueueTest.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Test
+    public void test_migrateTaskCount_decremented_onClear() {
```
most test in HZ have the following naming convention:

nameMethod_whenFoo_thenBla.

This makes it very easy to figure out where the test of the method is found and which cases are being tested. 

In this case I would expect the method to be called clear_whenMigrationTaskQueued_thenMigrationCountDecreased

metanet replied:
This is my preference. Unless there is a strict rule against my preference, i wont change it.

pveentjer commented on hazelcast/src/test/java/com/hazelcast/partition/impl/MigrationQueueTest.java:
```diff
@@ -0,0 +1,77 @@
+    @Test
+    public void test_migrateTaskCount_decremented_onClear() {
+        migrationQueue.add(mock(InternalPartitionServiceImpl.MigrateTask.class));
+        migrationQueue.clear();
```
Also add a line of white space between your setup and your action. Even though your test methods are very small, it still helps to see the general structure faster.

metanet replied:
No, i won't. Test is already too short

pveentjer commented on hazelcast/src/main/java/com/hazelcast/partition/impl/MigrationQueue.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+class MigrationQueue {
```
Can you add a few lines of javadoc for other engineers what the purpose is of this class?

metanet commented:
I am done with this pr. I wont make any change related to its style.

pveentjer commented:
There are some other comments that need to be addressed like a few lines  of Javadoc. Just to give the reader a quick understanding of the purpose (providing info of a migration is pending/executing) of the class (don't care about the methods).

metanet commented:
I added Javadoc for ``MigrationQueue`` class. I also updated commit message and pr description to make purpose of the pr more understandable. Lastly, I answered your concern about the callback in my reply above.

pveentjer commented:
Good.

pveentjer commented:
:+1:
